smart-contracts/contracts/mixins/MixinPurchase.sol

```diff
@@ -84,21 +84,20 @@ contract MixinPurchase is
       netPrice = inMemoryKeyPrice - discount;
     }
 
-    // We explicitly allow for greater amounts of ETH to allow 'donations'
-    _chargeAtLeast(netPrice);
-
     // Assign the key
     Key storage toKey = _getKeyFor(_recipient);
     uint previousExpiration = toKey.expirationTimestamp;
-    if (toKey.tokenId == 0) {
-      _assignNewTokenId(toKey);
-      _recordOwner(_recipient, toKey.tokenId);
+    if (toKey.expirationTimestamp >= block.timestamp) {
+      // This is an existing owner trying to extend their key
+      toKey.expirationTimestamp = previousExpiration.add(expirationDuration);
+    } else {
+      if (toKey.tokenId == 0) {
+        _assignNewTokenId(toKey);
+        _recordOwner(_recipient, toKey.tokenId);
+      }
       // SafeAdd is not required here since expirationDuration is capped to a tiny value
       // (relative to the size of a uint)
       toKey.expirationTimestamp = block.timestamp + expirationDuration;
```

Comment:
won't this line be executed even if line #92 is also executed?

Reply:
it seems like we want either line 92 or line 100 to be executed...not both.

Reply:
Na, we are nested inside an `else` block here.

Reply:
yup.